docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+These migrations always begin at the `canonical` version, and settings extensions are free to perform the migrations however they wish; however, Bottlerocket's Settings SDK will provide utilities to help settings developers support a series of linear setting upgrades and downgrades.
+Read on for more information about the Settings SDK
+
+All migrations are performed and stored at setting-write-time because Bottlerocket rollbacks may result in running an OS with a settings extension which does not support the version at which a given setting is stored.
+If Bottlerocket boots and finds that a setting is stored at a value which is not supported by it's installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent dataloss in the case that the image can eventually update to a new image which supports the stored setting version.
+
+### Configuration Templating
+Once Bottlerocket settings are stored in the datastore, they must somehow then be used to influence the behavior of the system.
+Fetching settings from the Bottlerocket API/datastore is currently done via two mechanisms:
+* Fetching data directly from the API using `apiclient`
+* Rendering config files (or just strings) using `schnauzer`
+
+For open-source tools, Bottlerocket prefers the rendered config file approach, as it doesn't require patching upstream code in order to consume the Bottlerocket API.
+First-party tools have no such constraints, and so today there exists  tools that take either approach.
+
+Because settings are discovered at runtime on the system, and indeed may no longer be handled by models written in Rust, it will no longer be feasible to create an `apiclient` with the same strong typing as exists today.
+It will instead be preferred that system services, including first-party services, all route their fetching of API settings through their own configuration files, which will be rendered by Bottlerocket whenever the related settings are changed.
+
+Today, configuration templates are rendered by `schnauzer` using [handlebars](https://handlebarsjs.com/) templates.
+`schnauzer` fetches the entirety of the settings tree, defines some useful helper functions, and then uses these as context for rendering the template.
+With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
+These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `+++` as such:
+
+```toml
++++
+[required-extensions]
+hostname = v1
+kubernetes = v1
++++
```

Comment:
I still have some doubts about the "front matter" approach, but I'm willing to defer judgment until I see the implementation 😀 

I'm still somewhat partial to the filesystem-oriented approach, e.g.:
```
usr/
└── share
    └── templates
        └── etc
            └── kubelet
                ├── kubelet.config
                └── kubelet.config.required.d
                    ├── kubernetes
                    └── network
```

Since this is a bit more consistent with some of the other ways dependencies are modeled, at the cost of giving up tight binding to the template.

Reply:
I'll open another PR discussing particulars about this design.